### PR TITLE
(PUP-8199) Use instances of Task instead of different Task sub-types

### PR DESCRIPTION
**Pull request by thallgren**
Prior to this commit, each task that was described with metadata became
a type of its own in the puppet type system. An instance of such a task
would represent a task invocation, complete with settings and parameter
values. This was problematic, primarily because the parameters were
represented by attributes and hence subject to the same naming rules as
the attributes, but also because it was cumbersome to filter out tasks
from other types in the type system.

This commit changes the way the `Task` type is used so that a task
description, instead of becoming a type in its own right, becomes an
instance of the one and only `Task` type. Creating an object that
represents a task invocation (description + parameter values) is no
longer possible.

The loader namespace `:task` is added to the Loaders. This namespace
must be used when searching for tasks. Using `:type` will now only find
the `Task` type.

### CODE REVIEW

File: lib/puppet_pal.rb [hunk 1]
```diff
@@ -389,8 +362,30 @@ def initialize(task_type)
...
+      params = @task.parameters
+      params_type = if params.nil?
+        T_GENERIC_TASK_HASH
+      else
+        key_to_type = {}
+        @task.parameters.each_pair { |k, v| key_to_type[k] = v['type'] }
+        Puppet::Pops::Types::TypeFactory.struct(key_to_type)
+      end
+      params_type.instance?(args_hash)
+    end
```
Comment (hlindberg) [inline lib/puppet_pal.rb]: I forgot to add  the same support for calling a lambda with a type mismatch error message as in the callable_with? for functions and plans. The intention was that `runnable_with?` should do the same. Not sure if that alters the approach to type checking, and/or if that should be done here, or in a separate ticket.

File: spec/unit/pops/types/task_spec.rb [hunk 2]
```diff
@@ -69,50 +69,24 @@ def compile(code = nil)
           }
         }
 
-        it 'loads task as a GenericTask subtype' do
+        it 'loads task without as a generic Task' do
```
Comment (hlindberg) [inline spec/unit/pops/types/task_spec.rb]: without what?

File: spec/unit/pops/types/task_spec.rb [hunk 3]
```diff
@@ -246,6 +246,25 @@ def compile(code = nil)
...
+          }
+
+          it 'loads the task with parameters set to undef' do
+            compile do
+              module_loader = Puppet.lookup(:loaders).find_loader('testmodule')
+              task = module_loader.load(:task, 'testmodule::hello')
+              expect(task_t.instance?(task)).to be_truthy
+              expect(task.parameters).to be_nil
+            end
+          end
```
Comment (hlindberg) [inline spec/unit/pops/types/task_spec.rb]: test missing for the semantics of having no parameters defined. Does that mean that "anything goes", or that no parameters are allowed? (I think it is "anything goes").

Comment (thallgren) [reply spec/unit/pops/types/task_spec.rb]: Yes, anything goes when parameters are not defined. I'll add a test.

Comment (thallgren) [reply spec/unit/pops/types/task_spec.rb]: Actually, the fact that no parameters means "anything goes" is implemented in the {{Puppet::Pal::TaskSignature}}, and there is a test for it. The task itself has no opinion about this.

**Pull request merged**